### ISSUE TITLE
CARLA setup

### ISSUE THREAD

**Issue by dylan-gonzalez** [Monash-Connected-Autonomous-Vehicle/autoware#17]
CARLA Debian package only has support for Ubuntu 18.04 and Ubuntu 22.04

- [x] I want a setup for CARLA that works regardless of host OS. Follow [docker installation](https://carla.readthedocs.io/en/latest/build_docker/)
- [ ] Get Autoware Universe working with CARLA. See [these tutorials](https://www.youtube.com/watch?v=EFH-vVxn180) made by Hatem Darweesh. I believe he's made his own 'OpenPlanner' thing to get this to work, not too sure... find out what exactly it is.
- [ ] Get the Twizy working with CARLA for Hardware-in-the-loop (HIL) demo

**Comment by Jokua** [comment]
when testing out the tutorial dylan posted, there will be a problem related to no octomap cmake or the like, so it can be solved by running ```sudo apt install ros-humble-octomap-ros```

![Image](https://github.com/Monash-Connected-Autonomous-Vehicle/autoware/assets/47382093/9544a5e3-befb-4347-85ff-98d2a136ccbd)

**Comment by dylan-gonzalez** [comment]
does the OpenPlanner tutorial use Ros humble?

**Comment by Jokua** [comment]
ran into another problem when following the instructions in the carla-autoware demo



![Image](https://github.com/Monash-Connected-Autonomous-Vehicle/autoware/assets/47382093/146847c2-faf7-4dd9-982e-c49722c9bf47)



![Image](https://github.com/Monash-Connected-Autonomous-Vehicle/autoware/assets/47382093/6b6e22f3-3088-4e91-ab69-8c5ce6d76406)

**Comment by Jokua** [comment]
going to continue seeing if i can do that, but otherwise the zenoh carla thing dylan posted might be another option because i found the github page for it https://github.com/evshary/autoware_carla_launch

**Comment by dylan-gonzalez** [comment]
so no errors so far??

**Comment by Jokua** [comment]
ok, just now got the bridge built, no errors so far



![Image](https://github.com/Monash-Connected-Autonomous-Vehicle/autoware/assets/47382093/5d4b53d4-b760-40d3-846a-2fb35b974d2f)

Don't panic if the folder location doesn't look right, what i understood what rocker does is that it "loads" the folders from your host machine into the docker container, and even if it gets removed every change to those folders stays so you don't lose progress (i think so far)

**Comment by Jokua** [comment]
building the autoware packages in another container now (fingers crossed it works bcs i didn't bother checking if it uses localhost or not but eh we'll cross that bridge when it gets to it). It seems to be downloading a custom Zenoh Version of autoware, so going to take a while.



![Image](https://github.com/Monash-Connected-Autonomous-Vehicle/autoware/assets/47382093/57e23508-32c6-44db-8c94-c77cab282842)

**Comment by Jokua** [comment]
just finished building the autoware packages

![Image](https://github.com/Monash-Connected-Autonomous-Vehicle/autoware/assets/47382093/3c8a37df-9882-488d-ad3f-e9de4e64f514)

**Comment by Jokua** [comment]
both containers are now built, now only left to test the demo case to see if it works

**Comment by Jokua** [comment]
ok, when running these commands for the demo, the map for carla did change, so at least that's working for now

![Image](https://github.com/Monash-Connected-Autonomous-Vehicle/autoware/assets/47382093/3ae5fd5a-6eeb-4c25-91a6-698b5d7c0586)

**Comment by Jokua** [comment]
but i'm not particularly sure when it's asking for the host1 ip (the pc running carla ) is the docker IP OR the actual host IP

**Comment by Jokua** [comment]
well seems like it works, i took the docker IP after finding it using ```hostname -I```


![Image](https://github.com/Monash-Connected-Autonomous-Vehicle/autoware/assets/47382093/f40ec7af-097d-419f-abe2-7332450450cc)


![Image](https://github.com/Monash-Connected-Autonomous-Vehicle/autoware/assets/47382093/0bf20562-6882-475f-83e1-f8d6be794e5c)

the IP address that starts with 172

**Comment by Jokua** [comment]
ok, seems like i can get the car to move by setting the goal pose, but dang it is LAGGY


![Image](https://github.com/Monash-Connected-Autonomous-Vehicle/autoware/assets/47382093/868214ec-da26-4950-a858-69e7f56916e3)

**Comment by Jokua** [comment]
yep, it works, and confirmed on both the rviz side and CARLA sim side, but it's really laggy so running in headless mode for CARLA should be more sensible.

**Comment by dylan-gonzalez** [comment]
11-12-23

@Jokua to do some more testing with CARLA + Autoware-Universe

- try some parking scenarios
- update the [wiki](https://github.com/Monash-Connected-Autonomous-Vehicle/autoware/wiki/Autoware) with CARLA launch params
- try multiple autoware vehicles?? (since it uses that zenoh bridge thing)

**Comment by Jokua** [comment]
ok, a bit of an update on what I have done so far:

I have managed to reduce the burden on the pc by running CARLA in "headless mode", which means it's gonna run without the GUI so it saves us some more resources.


![Image](https://github.com/Monash-Connected-Autonomous-Vehicle/autoware/assets/47382093/0a276794-6211-42a2-a843-2a8b445e3a0d)

it can be done when creating a new docker container with this command ```docker run  -p 2000-2002:2000-2002  --cpuset-cpus="0-5"  --runtime=nvidia  --gpus 'all,"capabilities=graphics,utility,display,video,compute"'  -e DISPLAY=$DISPLAY  -v /tmp/.X11-unix:/tmp/.X11-unix  -it  carlasim/carla  ./CarlaUE4.sh -opengl $1 -RenderOffScreen -quality-level=Epic```

Also i've managed to make it run properly on my laptop (well once, will update if i run into any issues because there was some error that goes like ```sh: 1: xdg-user-dir: not found``` when trying to follow the tutorial for setting up carla in your local env in [this tutorial.](https://antc2lt.medium.com/carla-on-ubuntu-20-04-with-docker-5c2ccdfe2f71#:~:text=Setting%20up%20the%20local%20environment)

**Comment by Jokua** [comment]
I have managed to run two instances as the demo has shown



![Image](https://github.com/Monash-Connected-Autonomous-Vehicle/autoware/assets/47382093/4796980b-49f4-4c9e-a7d4-1cc4f3e4843b)

I have also managed to get them both to move at the same time, though it is sort of laggy?


![Image](https://github.com/Monash-Connected-Autonomous-Vehicle/autoware/assets/47382093/6e4b8f1c-c32d-4936-9364-ef482b5c94ad)

Also I don't know why it decided to stop running by itself after moving once, so that's probably another thing to make note of for troubleshooting later.

Additionally, there is a FAQ on the Zenoh's github page about how to change the map [here](https://github.com/evshary/autoware_carla_launch/blob/main/carla_map/README.md)

Still unsure on how to change the car's URDF or model, might have to look up on how carla spawns it or ask someone who has used CARLA before.

**Comment by dylan-gonzalez** [comment]
> I have managed to reduce the burden on the pc by running CARLA in "headless mode", which means it's gonna run without the GUI so it saves us some more resources.

This would be very useful for the CI/CD pipeline guys @AnthonyZhOon @sxppro @npnquang 

> I have also managed to get them both to move at the same time, though it is sort of laggy?

If you record a rosbag and play it back again, is it still laggy?

**Comment by Jokua** [comment]
>If you record a rosbag and play it back again, is it still laggy?

I had to look up what a rosbag is, but how would I do that exactly?

Also I'll update the wiki on how to set it up and use it in the beast as well, since I've confirmed it can work on my laptop (be warned it takes ages if your internet is shite to download the zenoh autoware bridge and autoware docker images)

**Comment by dylan-gonzalez** [comment]
@Jokua if you're interested https://github.com/orgs/autowarefoundation/discussions/2828#discussioncomment-7929757

**Comment by Jokua** [comment]
oh, thanks! I'll check it out, hopefully the guy updated some things

**Comment by dylan-gonzalez** [comment]
@Jokua to do some more testing for the next 2 days

**Comment by Jokua** [comment]
I have just finished some testing, seems like the problem I had about not being able to set another route may be just slight user error, apart from that recorded quite a lengthy video (with some cropping issues, i accidentally captures my other screen as well, as well as the fact i couldn't record my own voice). But otherwise, the basic functionality seems to be working for a single car RViz running, i spawned two cars but only operated one due to my laptop spec contraints.
